core/services/ocr3/plugins/ccip/internal/reader/home_chain.go

```
@@ -0,0 +1,152 @@
+package reader
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	mapset "github.com/deckarep/golang-set/v2"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+)
+
+type HomeChainConfigPoller struct {
+	homeChainReader  types.ContractReader
+	homeChainConfig  cciptypes.HomeChainConfig
+	lggr             logger.Logger
+	mutex            sync.RWMutex
+	backgroundCancel context.CancelFunc
+}
+
+func NewHomeChainConfigPoller(
+	homeChainReader types.ContractReader,
+	lggr logger.Logger,
+) *HomeChainConfigPoller {
+	return &HomeChainConfigPoller{
+		homeChainReader: homeChainReader,
+		lggr:            lggr,
+	}
+}
+
+func (r *HomeChainConfigPoller) Start(ctx context.Context) error {
+
+	r.mutex.Lock()
+	if r.backgroundCancel != nil {
+		r.lggr.Errorw("Polling already started")
+		// We don't want to return an actual error here as it's already working as expected
```

You can return an error, client should handle it

---

You never unlock, this will lead to big issues

---

good catch!